feat: `rawdb.InspectDatabaseOption`

ARR4N opened the pull request:
## Why this should be merged

Allows for `ava-labs/coreth` equivalent modifications of `rawdb.InspectDatabase()` through external logic injection.

## How this works

Variadic options to:

1. Record a database statistic;
2. Mark a database statistic as metadata;
3. Filter statistics for printing.

## How this was tested

Testable example acting as a unit test.

ARR4N committed:
feat: `rawdb.InspectDatabaseOption`

ARR4N committed:
test: `DatabaseStat` example output

qdm12 commented on core/rawdb/database.libevm.go:
```diff
@@ -0,0 +1,89 @@
+		c.statRecorder = rec
+	})
+}
+
+// A DatabaseStat stores total size and counts for a parameter measured by
+// [InspectDatabase]. It is exported for use with [WithDatabaseStatRecorder].
+type DatabaseStat = stat
```
I think this can be removed?

```suggestion
```

The test `ExampleDatabaseStat` can be removed as well I think

ARR4N replied:
Any use of `WithDatabaseStatRecorder` needs to use it. Yes, it can be reimplemented but it's literally the identical code for the identical purpose so it shouldn't be repeated.

qdm12 commented on core/rawdb/database.libevm.go:
```diff
@@ -0,0 +1,89 @@
+// for every `key` not otherwise matched by the [InspectDatabase] iterator loop.
+// The returned boolean signals whether the recorder matches the key, thus
+// stopping further matches.
+func WithDatabaseStatRecorder(rec func(key []byte, _ common.StorageSize) bool) InspectDatabaseOption {
```
nit let's name the 2nd argument? Otherwise it may sound like it's never used.


```suggestion
func WithDatabaseStatRecorder(rec func(key []byte, size common.StorageSize) bool) InspectDatabaseOption {
```

qdm12 commented on core/rawdb/database.go:
```diff
@@ -451,7 +452,8 @@ func (s *stat) Count() string {
 // InspectDatabase traverses the entire database and checks the size
 // of all different categories of data.
-func InspectDatabase(db ethdb.Database, keyPrefix, keyStart []byte) error {
+func InspectDatabase(db ethdb.Database, keyPrefix, keyStart []byte, opts ...InspectDatabaseOption) error {
```
It might be interesting to write a test/example-test using `InspectDatabase` with options. Aka this really feels like we need a test for InspectDatabase (both here and upstream)

ARR4N replied:
An example would be great because it also avoids having to change the `os.Stdout` destination to make testing possible.

qdm12 replied:
I can work on this, let me know 😉 And yes an example fits perfectly (and we can PR upstream as well)

darioush reviewed (commented):
Let's add a test for the new behavior and remove WithDatabaseStatRecorder if it is unused. Then LGTM

ARR4N committed:
Apply suggestions from code review

Co-authored-by: Quentin McGaw <quentin.mcgaw@avalabs.org>
Signed-off-by: Arran Schlosberg <519948+ARR4N@users.noreply.github.com>

qdm12 committed:
chore(core/rawdb): `ExampleInspectDatabase` test (#119)

## Why this should be merged

Add an example test for `InspectDatabase` with options.

## How this works

See example code

ARR4N committed:
refactor: remove `-1` from `stubIterator` literal

qdm12 commented on core/rawdb/database.libevm_test.go:
```diff
@@ -0,0 +1,193 @@
+func (s *stubIterator) Next() bool {
+	s.i++
+	available := s.pos() < len(s.kvs)
+	return available
```
nit saving a line - if `available` feels necessary, we should have it as a named return `Next() (available bool)`, but I feel it's fine without 

```suggestion
	return s.pos() < len(s.kvs)
```

ARR4N replied:
Agreed; I wanted that anyway but figured you liked the verbosity.

qdm12 commented on core/rawdb/database.libevm_test.go:
```diff
@@ -0,0 +1,193 @@
+
+type stubIterator struct {
+	ethdb.Iterator
+	i   int // see pos()
```
nit is it possible to use square brackets on `pos`? i.e. `// see [stubIterator.pos]`?

qdm12 commented on core/rawdb/database.libevm_test.go:
```diff
@@ -0,0 +1,193 @@
+	return nil, nil
+}
+
+func (s *stubDatabase) ReadAncients(fn func(ethdb.AncientReaderOp) error) (err error) {
```
nit unneeded named return

```suggestion
func (s *stubDatabase) ReadAncients(fn func(ethdb.AncientReaderOp) error) error {
```

ARR4N replied:
You can't merge a PR into my branch only to nit-pick your own code 🤣 haha, do that first 😛

qdm12 commented on core/rawdb/database.libevm_test.go:
```diff
@@ -0,0 +1,193 @@
+				default:
+					modified = append(modified, line)
+				}
+			}
```
nit Should we remove this block perhaps?
I suppose it's useful for coreth and subnet-evm, but maybe making the example a bit too long?

ARR4N replied:
Now that you've added the comment to `WithDatabaseStatsTransformer` to explain what each line is, I agree that there's no point showing how to delete lines. Removed.

qdm12 commented on core/rawdb/database.libevm_test.go:
```diff
@@ -0,0 +1,193 @@
+		rawdb.WithDatabaseMetadataKeys(func(key []byte) bool {
+			return bytes.Equal(key, []byte("mymetadatakey"))
+		}),
+		rawdb.WithDatabaseStatsTransformer(func(s [][]string) [][]string {
```
can we call the argument `lines` instead of `s`? Or `rows` maybe is more appropriate? `s` is rather not very helpful 🤔

qdm12 commented on core/rawdb/database.libevm_test.go:
```diff
@@ -0,0 +1,193 @@
+
+	// Output:
+	// Sum: 7.00 B
+	// Count: 4
```
The 2 last adds don't add (pun intended) anything to the example I think

```suggestion
	stat.Add(common.StorageSize(1)) // only to demonstrate param type
	stat.Add(2)

	fmt.Println("Sum:", stat.Size())    // sum of all values passed to Add()
	fmt.Println("Count:", stat.Count()) // number of calls to Add()

	// Output:
	// Sum: 3.00 B
	// Count: 2
```

qdm12 commented on core/rawdb/database.libevm.go:
```diff
@@ -0,0 +1,89 @@
+// WithDatabaseStatsTransformer returns an option that causes all statistics to
+// be passed to the provided function, with its return value being printed
+// instead of the original values.
+func WithDatabaseStatsTransformer(transform func([][]string) [][]string) InspectDatabaseOption {
```
nit for documentation purposes

```suggestion
func WithDatabaseStatsTransformer(transform func(rows [][]string) [][]string) InspectDatabaseOption {
```

qdm12 commented on core/rawdb/database.libevm.go:
```diff
@@ -0,0 +1,89 @@
+
+// WithDatabaseStatsTransformer returns an option that causes all statistics to
+// be passed to the provided function, with its return value being printed
+// instead of the original values.
```
```suggestion
// WithDatabaseStatsTransformer returns an option that causes all statistics rows to
// be passed to the provided function, with its return value being printed
// instead of the original values.
// Each row contains 4 columns: database, category, size and count.
```

qdm12 reviewed (approved):
Pre-approving to get this merged 🛥️ !

ARR4N committed:
Apply suggestions from code review

Co-authored-by: Quentin McGaw <quentin.mcgaw@avalabs.org>
Signed-off-by: Arran Schlosberg <519948+ARR4N@users.noreply.github.com>

ARR4N committed:
refactor: implement review requests
